### PR TITLE
Only run filter once.

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/58329

This PR replaces https://github.com/WordPress/wordpress-develop/pull/4874. Basically, only run the `filter` method once in `get_term`. Run if the object has been filtered or the filter parameter does match the one passed in the function. 

This fix would also fix the issues found in  https://core.trac.wordpress.org/ticket/58327

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/term.php
Comment (mukeshpanchal27) [inline tests/phpunit/tests/term.php]:
```suggestion
	 * @ticket 58329
	 *
	 * @dataProvider data_get_term_filter
	 *
	 * @param string $filter Term to filter.
```

Update `$filter` parameter description if it not correct.

File: tests/phpunit/tests/term.php
Comment (mukeshpanchal27) [inline tests/phpunit/tests/term.php]:
```suggestion
	 * @ticket 58329
	 *
	 * @dataProvider data_get_term_filter
	 *
	 * @param string $filter Term to filter.
```

File: tests/phpunit/tests/term.php
Comment (mukeshpanchal27) [inline tests/phpunit/tests/term.php]:
```suggestion
			static function () use ( $cat ) {
```

Per [Changeset 56559](https://core.trac.wordpress.org/changeset/56559), Include one space after function keyword for closures and use `static`

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @spacedmonkey for the PR. Left some unit test related feedbacks.

**Comment by spacedmonkey** [comment]
> Thanks @spacedmonkey for the PR. Left some unit test related feedbacks.

Actioned feedback.

File: tests/phpunit/tests/term.php
Comment (costdev) [inline tests/phpunit/tests/term.php]:
```suggestion
		$this->assertSame( $filter, $term->filter, "The term's 'filter' property should be set to '$filter'." );
```
Some added clarity if the assertion fails without increasing the maintenance burden.

image

with `--testdox`:

image

File: tests/phpunit/tests/term.php
Comment (costdev) [inline tests/phpunit/tests/term.php]:
```suggestion
		$this->assertSame( $filter, $term->filter, "The term's 'filter' property should be set to '$filter'." );
```
Same as [this comment](https://github.com/WordPress/wordpress-develop/pull/5260/files#r1331575257).

File: tests/phpunit/tests/term.php
Comment (costdev) [inline tests/phpunit/tests/term.php]:
```suggestion
	 * @param string $filter How to sanitize term fields.
```

File: tests/phpunit/tests/term.php
Comment (costdev) [inline tests/phpunit/tests/term.php]:
```suggestion
	 * @param string $filter How to sanitize term fields.
```

File: tests/phpunit/tests/term.php
Comment (costdev) [inline tests/phpunit/tests/term.php]:
```suggestion
	public function test_get_term_should_set_term_filter_property_to_filter_argument( $filter ) {
```
For clarity about what's being tested.

File: tests/phpunit/tests/term.php
Comment (joemcgill) [inline tests/phpunit/tests/term.php:311]: Remove extra space.

```suggestion
```

**Review by joemcgill — APPROVED** [review]
I like this approach much better. The tests that you've added already pass without this change in `trunk`, so it would be nice to add a test that demonstrates the exact problem you're trying to fix. I've been trying to work something up, but haven't had success yet. Will keep messing with it, but you may beat me to it.

**Comment by spacedmonkey** [comment]
> I like this approach much better. The tests that you've added already pass without this change in `trunk`, so it would be nice to add a test that demonstrates the exact problem you're trying to fix. I've been trying to work something up, but haven't had success yet. Will keep messing with it, but you may beat me to it.

@joemcgill The fact that the test works in trunk was the point. To see no breakage. I have added one more unit test, that proved that sanativation is only run once. Can you take another look at this PR.

File: tests/phpunit/tests/term.php
Comment (costdev) [inline tests/phpunit/tests/term.php]:
```suggestion
	/**
	 * @ticket 58329
	 *
	 * @covers ::get_term
	 *
	 * @dataProvider data_get_term_filter
	 *
	 * @param string $filter How to sanitize term fields.
	 */
```
Let's add `@covers ::get_term` annotations to the tests.

File: tests/phpunit/tests/term.php
Comment (costdev) [inline tests/phpunit/tests/term.php]:
```suggestion
		$this->assertSame( 0, $filter->get_call_count(), 'The term was filtered more than once' );
```

**Review by joemcgill — APPROVED** [review]
Thanks @spacedmonkey! I understood why you had written the original tests. I just wanted to make sure that we also had a test that confirmed the filter is only being run once (so a future change doesn't cause a regression). The new test you've added seems to do the trick!

**Review by costdev — COMMENTED** [review]
Thanks for the updates @spacedmonkey! Just a few final notes :slightly_smiling_face:

File: tests/phpunit/tests/term.php
Comment (spacedmonkey) [inline tests/phpunit/tests/term.php:349]:
```suggestion
	 * @ticket 58329
	 *
	 * @covers ::get_term
```

File: tests/phpunit/tests/term.php
Comment (spacedmonkey) [inline tests/phpunit/tests/term.php:313]:
```suggestion
	 * @ticket 58329
	 *
	 * @covers ::get_term
```

**Comment by spacedmonkey** [comment]
> Thanks for the updates @spacedmonkey! Just a few final notes 🙂

Thanks @costdev Feedback actioned. Once I get an approval from you, I will commit.

**Review by costdev — APPROVED** [review]
Thanks @spacedmonkey! LGTM :+1:

**Review by mukeshpanchal27 — APPROVED** [review]

**Comment by spacedmonkey** [comment]
Committed https://github.com/WordPress/wordpress-develop/commit/c1247f6a278bd6a5fec799964177743b5d26a2aa

**Pull request closed without merging**
